src/machine/image-varlink.c

```diff
@@ -277,3 +280,130 @@ int vl_method_set_pool_limit(sd_varlink *link, sd_json_variant *parameters, sd_v
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(name);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("name", name),
+                        JSON_BUILD_PAIR_UNSIGNED_NOT_EQUAL("usageExclusive", usage_exclusive, UINT64_MAX));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+static int clean_pool_done_internal(Operation *operation, FILE *file, int child_error) {
+        int r;
+
+        assert(operation);
+        assert(operation->link);
+
+        r = clean_pool_read_first_entry(file, child_error, /* error = */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to read first entry from tmp file: %m");
+
+        /* On success the resulting temporary file will contain a list of image names that were removed followed by
+         * their size on disk. Let's read that and turn it into a bus message. */
+        _cleanup_free_ char *previous_name = NULL;
+        uint64_t previous_usage;
+        for (;;) {
+                _cleanup_free_ char *name = NULL;
+                uint64_t usage;
+                r = clean_pool_read_next_entry(file, &name, &usage);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read next entry from tmp file: %m");
+                if (r == 0)
+                        break;
+
+                if (previous_name) {
+                        r = clean_pool_list_one_image(operation->link, previous_name, previous_usage, /* more = */ true);
+                        if (r < 0)
+                                return r;
+                }
+
+                previous_name = TAKE_PTR(name);
+                previous_usage = usage;
+        }
+
+        if (previous_name)
+                return clean_pool_list_one_image(operation->link, previous_name, previous_usage, /* more = */ false);
+
+        // (XXX) return sd_varlink_reply(operation->link, NULL); ????
+        return sd_varlink_error(operation->link, "io.systemd.MachineImage.NoSuchImage", NULL);
+}
+
+static int clean_pool_done(Operation *operation, int child_error, sd_bus_error *error) {
+        _cleanup_fclose_ FILE *file = NULL;
+        int r;
+
+        assert(operation);
+        assert(operation->link);
+        assert(operation->extra_fd >= 0);
+
+        file = take_fdopen(&operation->extra_fd, "r");
+        if (!file)
+                return log_debug_errno(errno, "Failed to take opened tmp file's fd: %m");
+
+        r = clean_pool_done_internal(operation, file, child_error);
+        if (r < 0)
+                r = sd_varlink_error_errno(operation->link, r);
+        if (r < 0)
+                log_debug_errno(r, "Failed to reply to varlink request, ignoring: %m");
```

Comment:
i think these two lines should be inside an {} of the `if (r < 0)` block before. because they really just do error handling for the sd_varlink_error_errno(), not for th eclean_pool_done_internal())